File: stack/min_stack/min_stack.cc
#include <stack>

using namespace std;

class MinStack {
private:
    stack<int> stck;
    stack<int> min;

public:
    MinStack() {
        
    }
    
    void push(int val) {
        stck.push(val);
        if (min.empty() || val < min.top()) {
            min.push(val);
        } else {
            min.push(min.top());
        }
    }
    
    void pop() {
        stck.pop();
        min.pop();
    }
    
    int top() {
        return stck.top();
    }
    
    int getMin() {
        return min.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: stack/min_stack/min_stack.java
import java.util.Stack;

class MinStack {
  private Stack<Integer> stack;
  private Stack<Integer> min;

  public MinStack() {
      stack = new Stack<Integer>();
      min = new Stack<Integer>();
  }
  
  public void push(int val) {
      stack.push(val);
      min.push(Math.min(val, min.empty() ? val : min.peek()));
  }
  
  public void pop() {
      stack.pop();
      min.pop();
  }
  
  public int top() {
      return stack.peek();
  }
  
  public int getMin() {
      return min.peek();
  }
}